File: backend/FinanceTracker.Application/Services/TransactionCsvService.cs
using System.Globalization;
using System.Text;
using FinanceTracker.Application.Abstractions;
using FinanceTracker.Application.Common;
using FinanceTracker.Application.Dtos;
using FinanceTracker.Domain;
using Microsoft.EntityFrameworkCore;

namespace FinanceTracker.Application.Services;

/// <summary>CSV export and import for transactions.</summary>
public sealed class TransactionCsvService(IAppDbContext db)
{
    public const string DateFormat = "yyyy-MM-dd";

    private const char TagDelimiter = ';';

    private static readonly string[] Header =
        ["Date", "Type", "Amount", "Account", "Category", "Description", "Notes", "Tags"];

    public async Task<string> ExportAsync(
        Guid userId,
        DateTime? from,
        DateTime? to,
        CancellationToken cancellationToken)
    {
        var query = db.Transactions.AsNoTracking().OwnedBy(userId);

        if (from.HasValue)
        {
            var fromUtc = UtcDate.Normalize(from.Value);
            query = query.Where(t => t.Date >= fromUtc);
        }

        if (to.HasValue)
        {
            var toUtc = UtcDate.Normalize(to.Value);
            query = query.Where(t => t.Date <= toUtc);
        }

        var transactions = await query
            .OrderByDescending(t => t.Date)
            .ThenByDescending(t => t.Id)
            .ToListAsync(cancellationToken)
            .ConfigureAwait(false);

        var accountNames = await LoadAccountNamesAsync(userId, cancellationToken).ConfigureAwait(false);
        var categoryNames = await LoadCategoryNamesAsync(userId, cancellationToken).ConfigureAwait(false);

        var builder = new StringBuilder();
        CsvFile.AppendRow(builder, Header);

        foreach (var t in transactions)
        {
            CsvFile.AppendRow(builder,
            [
                t.Date.ToString(DateFormat, CultureInfo.InvariantCulture),
                ToWireName(t.Type),
                t.Amount.ToString("0.00", CultureInfo.InvariantCulture),
                Lookup(accountNames, t.AccountId),
                t.CategoryId is { } categoryId ? Lookup(categoryNames, categoryId) : string.Empty,
                t.Description,
                t.Notes ?? string.Empty,
                string.Join(TagDelimiter, Transaction.SplitTags(t.TagsRaw)),
            ]);
        }

        return builder.ToString();
    }

    public async Task<ImportResult> ImportAsync(Guid userId, Stream csv, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(csv);

        using var reader = new StreamReader(csv, Encoding.UTF8, detectEncodingFromByteOrderMarks: true, leaveOpen: true);
        var content = await reader.ReadToEndAsync(cancellationToken).ConfigureAwait(false);

        var rows = CsvFile.Parse(content);
        if (rows.Count == 0)
        {
            throw AppException.BadRequest("The uploaded file contains no rows.");
        }

        var accounts = await db.Accounts
            .AsNoTracking()
            .OwnedBy(userId)
            .ToDictionaryAsync(a => a.Name.ToUpperInvariant(), a => a.Id, cancellationToken)
            .ConfigureAwait(false);

        var categories = await CategoryService
            .VisibleTo(db.Categories.AsNoTracking(), userId)
            .ToListAsync(cancellationToken)
            .ConfigureAwait(false);

        var categoryIds = categories
            .GroupBy(c => c.Name.ToUpperInvariant())
            .ToDictionary(g => g.Key, g => g.First().Id);

        var imported = new List<Transaction>();
        var skipped = 0;

        foreach (var row in rows.Skip(IsHeaderRow(rows[0]) ? 1 : 0))
        {
            var transaction = TryBuild(userId, row, accounts, categoryIds);
            if (transaction is null)
            {
                skipped++;
            }
            else
            {
                imported.Add(transaction);
            }
        }

        if (imported.Count > 0)
        {
            db.Transactions.AddRange(imported);
            await db.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
        }

        return new ImportResult(imported.Count, skipped);
    }

    private static Transaction? TryBuild(
        Guid userId,
        List<string> row,
        Dictionary<string, Guid> accounts,
        Dictionary<string, Guid> categories)
    {
        if (row.Count < 6)
        {
            return null;
        }

        if (!DateTime.TryParse(
                Field(row, 0),
                CultureInfo.InvariantCulture,
                DateTimeStyles.AdjustToUniversal | DateTimeStyles.AssumeUniversal,
                out var date))
        {
            return null;
        }

        if (!TryParseType(Field(row, 1), out var type) || type == TransactionType.Transfer)
        {
            // Transfers need a destination account, which the CSV shape does not carry.
            return null;
        }

        if (!decimal.TryParse(Field(row, 2), NumberStyles.Currency, CultureInfo.InvariantCulture, out var amount)
            || amount <= 0m)
        {
            return null;
        }

        if (!accounts.TryGetValue(Field(row, 3).ToUpperInvariant(), out var accountId))
        {
            return null;
        }

        var categoryName = Field(row, 4).ToUpperInvariant();
        Guid? categoryId = categories.TryGetValue(categoryName, out var found) ? found : null;
        var notes = Field(row, 6);
        var tags = Field(row, 7).Split(TagDelimiter, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        return new Transaction
        {
            UserId = userId,
            AccountId = accountId,
            CategoryId = categoryId,
            Type = type,
            Amount = decimal.Round(amount, 2, MidpointRounding.AwayFromZero),
            Date = DateTime.SpecifyKind(date, DateTimeKind.Utc),
            Description = Field(row, 5),
            Notes = string.IsNullOrWhiteSpace(notes) ? null : notes,
            TagsRaw = Transaction.JoinTags(tags),
        };
    }

    private static bool IsHeaderRow(List<string> row) =>
        row.Count > 0 && string.Equals(row[0].Trim(), "Date", StringComparison.OrdinalIgnoreCase);

    private static bool TryParseType(string value, out TransactionType type) =>
        Enum.TryParse(value.Trim(), ignoreCase: true, out type) && Enum.IsDefined(type);

    private static string Field(List<string> row, int index) =>
        index < row.Count ? row[index].Trim() : string.Empty;

    private static string ToWireName(TransactionType type) =>
        char.ToLowerInvariant(type.ToString()[0]) + type.ToString()[1..];

    private static string Lookup(Dictionary<Guid, string> names, Guid id) =>
        names.TryGetValue(id, out var name) ? name : string.Empty;

    private async Task<Dictionary<Guid, string>> LoadAccountNamesAsync(Guid userId, CancellationToken cancellationToken) =>
        await db.Accounts
            .AsNoTracking()
            .OwnedBy(userId)
            .ToDictionaryAsync(a => a.Id, a => a.Name, cancellationToken)
            .ConfigureAwait(false);

    private async Task<Dictionary<Guid, string>> LoadCategoryNamesAsync(Guid userId, CancellationToken cancellationToken) =>
        await CategoryService
            .VisibleTo(db.Categories.AsNoTracking(), userId)
            .ToDictionaryAsync(c => c.Id, c => c.Name, cancellationToken)
            .ConfigureAwait(false);
}

File: backend-go/internal/application/transaction_csv_service.go
package application

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"github.com/pharaujo/finance/backend-go/internal/domain"
)

// CSVDateFormat is TransactionCsvService.DateFormat ("yyyy-MM-dd"), used both
// for the Date column and for the download's file name.
const CSVDateFormat = "2006-01-02"

// tagDelimiter is the ';' the CSV column packs tags with; the storage column
// uses U+001F instead.
const tagDelimiter = ";"

// emptyCSVMessage is the 400 an unparseable upload raises.
const emptyCSVMessage = "The uploaded file contains no rows."

// csvHeader is the header row export writes and import recognises.
var csvHeader = []string{"Date", "Type", "Amount", "Account", "Category", "Description", "Notes", "Tags"}

// csvDateLayouts are the forms DateTime.TryParse accepts from an invariant
// culture that this API also accepts. .NET's parser recognises a longer list of
// exotic spellings; these are the ones a real export or spreadsheet produces.
var csvDateLayouts = []string{
	CSVDateFormat,
	time.RFC3339,
	"2006-01-02T15:04:05",
	"2006-01-02 15:04:05",
	"01/02/2006",
	"1/2/2006",
	"01/02/2006 15:04:05",
	"1/2/2006 15:04:05",
	"1/2/2006 3:04:05 PM",
}

// TransactionCsvService is CSV export and import, the Go twin of
// FinanceTracker.Application.Services.TransactionCsvService.
type TransactionCsvService struct {
	transactions TransactionRepository
	accounts     AccountRepository
	categories   CategoryRepository

	newID func() uuid.UUID
}

// NewTransactionCsvService wires the service to its ports.
func NewTransactionCsvService(
	transactions TransactionRepository,
	accounts AccountRepository,
	categories CategoryRepository,
) *TransactionCsvService {
	return &TransactionCsvService{
		transactions: transactions,
		accounts:     accounts,
		categories:   categories,
		newID:        uuid.New,
	}
}

// ExportFileName is the name the download is offered under.
func ExportFileName(now time.Time) string {
	return "transactions-" + now.UTC().Format(CSVDateFormat) + ".csv"
}

// Export renders the caller's transactions in an inclusive date window, newest
// first, as one CSV document.
func (s *TransactionCsvService) Export(
	ctx context.Context,
	userID uuid.UUID,
	from, to *time.Time,
) (string, error) {
	transactions, err := s.transactions.ListRange(
		ctx, userID, domain.NormalizeUTCPtr(from), domain.NormalizeUTCPtr(to))
	if err != nil {
		return "", fmt.Errorf("application: listing transactions for export: %w", err)
	}

	accountNames, err := s.accountNames(ctx, userID)
	if err != nil {
		return "", err
	}
	categoryNames, err := s.categoryNames(ctx, userID)
	if err != nil {
		return "", err
	}

	var builder strings.Builder
	AppendCSVRow(&builder, csvHeader)

	for _, transaction := range transactions {
		category := ""
		if transaction.CategoryId != nil {
			category = categoryNames[*transaction.CategoryId]
		}
		notes := ""
		if transaction.Notes != nil {
			notes = *transaction.Notes
		}

		AppendCSVRow(&builder, []string{
			transaction.Date.UTC().Format(CSVDateFormat),
			transaction.Type.String(),
			transaction.Amount.StringFixed(domain.MoneyScale),
			accountNames[transaction.AccountId],
			category,
			transaction.Description,
			notes,
			strings.Join(domain.SplitTags(transaction.TagsRaw), tagDelimiter),
		})
	}

	return builder.String(), nil
}

// Import reads a CSV document and inserts the rows it can make sense of. A row
// it cannot use is skipped rather than failing the upload, which is why the
// result reports both counts.
func (s *TransactionCsvService) Import(
	ctx context.Context,
	userID uuid.UUID,
	content []byte,
) (ImportResult, error) {
	rows := ParseCSV(strings.TrimPrefix(string(content), "\ufeff"))
	if len(rows) == 0 {
		return ImportResult{}, domain.BadRequest(emptyCSVMessage)
	}

	accountIDs, err := s.accountIDsByName(ctx, userID)
	if err != nil {
		return ImportResult{}, err
	}
	categoryIDs, err := s.categoryIDsByName(ctx, userID)
	if err != nil {
		return ImportResult{}, err
	}

	if isHeaderRow(rows[0]) {
		rows = rows[1:]
	}

	imported := make([]domain.Transaction, 0, len(rows))
	skipped := 0
	for _, row := range rows {
		transaction, ok := s.build(userID, row, accountIDs, categoryIDs)
		if !ok {
			skipped++
			continue
		}
		imported = append(imported, transaction)
	}

	if len(imported) > 0 {
		if err := s.transactions.AddMany(ctx, imported); err != nil {
			return ImportResult{}, fmt.Errorf("application: inserting imported transactions: %w", err)
		}
	}

	return ImportResult{Imported: len(imported), Skipped: skipped}, nil
}

// build is TryBuild: every check that fails means "skip this row", never "fail
// the upload".
func (s *TransactionCsvService) build(
	userID uuid.UUID,
	row []string,
	accountIDs, categoryIDs map[string]uuid.UUID,
) (domain.Transaction, bool) {
	if len(row) < 6 {
		return domain.Transaction{}, false
	}

	date, ok := parseCSVDate(csvField(row, 0))
	if !ok {
		return domain.Transaction{}, false
	}

	transactionType, ok := domain.ParseTransactionType(csvField(row, 1))
	// Transfers need a destination account, which the CSV shape does not carry.
	if !ok || !transactionType.Valid() || transactionType == domain.TransactionTransfer {
		return domain.Transaction{}, false
	}

	amount, ok := parseCurrencyAmount(csvField(row, 2))
	if !ok || !amount.IsPositive() {
		return domain.Transaction{}, false
	}

	accountID, ok := accountIDs[strings.ToUpper(csvField(row, 3))]
	if !ok {
		return domain.Transaction{}, false
	}

	var categoryID *uuid.UUID
	if found, ok := categoryIDs[strings.ToUpper(csvField(row, 4))]; ok {
		categoryID = &found
	}

	notes := csvField(row, 6)
	tags := strings.Split(csvField(row, 7), tagDelimiter)

	return domain.Transaction{
		Id:          s.newID(),
		UserId:      userID,
		AccountId:   accountID,
		CategoryId:  categoryID,
		Type:        transactionType,
		Amount:      domain.RoundMoney(amount),
		Date:        date,
		Description: csvField(row, 5),
		Notes:       trimmedOrNil(&notes),
		TagsRaw:     domain.JoinTags(tags),
	}, true
}

func (s *TransactionCsvService) accountNames(
	ctx context.Context,
	userID uuid.UUID,
) (map[uuid.UUID]string, error) {
	accounts, err := s.accounts.List(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("application: listing accounts: %w", err)
	}

	names := make(map[uuid.UUID]string, len(accounts))
	for _, account := range accounts {
		names[account.Id] = account.Name
	}
	return names, nil
}

func (s *TransactionCsvService) categoryNames(
	ctx context.Context,
	userID uuid.UUID,
) (map[uuid.UUID]string, error) {
	categories, err := s.categories.ListVisible(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("application: listing categories: %w", err)
	}

	names := make(map[uuid.UUID]string, len(categories))
	for _, category := range categories {
		names[category.Id] = category.Name
	}
	return names, nil
}

// accountIDsByName keys accounts by upper-cased name, as the import lookup
// does. Two accounts sharing a name make the .NET dictionary throw; here the
// later row wins, so the upload still completes.
func (s *TransactionCsvService) accountIDsByName(
	ctx context.Context,
	userID uuid.UUID,
) (map[string]uuid.UUID, error) {
	accounts, err := s.accounts.List(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("application: listing accounts: %w", err)
	}

	ids := make(map[string]uuid.UUID, len(accounts))
	for _, account := range accounts {
		ids[strings.ToUpper(account.Name)] = account.Id
	}
	return ids, nil
}

// categoryIDsByName keeps the first category of each name, which is what the
// .NET GroupBy(...).First() does over the visible set.
func (s *TransactionCsvService) categoryIDsByName(
	ctx context.Context,
	userID uuid.UUID,
) (map[string]uuid.UUID, error) {
	categories, err := s.categories.ListVisible(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("application: listing categories: %w", err)
	}

	ids := make(map[string]uuid.UUID, len(categories))
	for _, category := range categories {
		key := strings.ToUpper(category.Name)
		if _, taken := ids[key]; !taken {
			ids[key] = category.Id
		}
	}
	return ids, nil
}

func isHeaderRow(row []string) bool {
	return len(row) > 0 && strings.EqualFold(strings.TrimSpace(row[0]), "Date")
}

// csvField reads one column, trimmed, treating a short row as blanks.
func csvField(row []string, index int) string {
	if index >= len(row) {
		return ""
	}
	return strings.TrimSpace(row[index])
}

// parseCSVDate mirrors DateTime.TryParse with AssumeUniversal: a value with no
// zone is read as UTC and one with an offset is converted.
func parseCSVDate(value string) (time.Time, bool) {
	for _, layout := range csvDateLayouts {
		if parsed, err := time.ParseInLocation(layout, value, time.UTC); err == nil {
			return parsed.UTC(), true
		}
	}
	return time.Time{}, false
}

// parseCurrencyAmount mirrors decimal.TryParse with NumberStyles.Currency:
// thousands separators, a leading currency symbol and parenthesised or trailing
// negatives are all accepted.
func parseCurrencyAmount(value string) (domain.Money, bool) {
	text := strings.TrimSpace(value)
	negative := false

	if strings.HasPrefix(text, "(") && strings.HasSuffix(text, ")") {
		negative = true
		text = text[1 : len(text)-1]
	}

	text = strings.TrimSpace(text)
	text = strings.TrimLeft(text, "¤$£€")
	text = strings.ReplaceAll(text, ",", "")
	text = strings.TrimSpace(text)

	if strings.HasSuffix(text, "-") {
		negative = true
		text = strings.TrimSuffix(text, "-")
	}

	amount, err := decimal.NewFromString(strings.TrimSpace(text))
	if err != nil {
		return domain.Zero(), false
	}
	if negative {
		amount = amount.Neg()
	}
	return amount, true
}
